packages/netsuite-adapter/src/client/client.ts

```diff
@@ -187,47 +230,86 @@ export default class NetsuiteClient {
-  ):Promise<Map<string, Set<string>>> {
-    const dependencyMap = new DefaultMap<string, Set<string>>(() => new Set())
-    const elemIdsAndCustInfoArr = await awu(changes)
-      .map(getChangeData)
-      .map(async element => ({
-        elemId: element.elemID,
-        custInfos: await NetsuiteClient.toCustomizationInfos(
-          [element], elementsSourceIndex
+  ): Promise<SDFObjectNode[]> {
+    const elementsAndChangeTypes = changes
+      .filter(isAdditionOrModificationChange)
+      .map(change =>
+        ({ element: getChangeData(change), changeType: getChangeType(change) }))
+    const elemIdSet = new Set(elementsAndChangeTypes.map(elementsAndChangeType =>
+      elementsAndChangeType.element.elemID.getFullName()))
+    return awu(elementsAndChangeTypes)
+      .filter(elementsAndChangeType => !isField(elementsAndChangeType.element)
+      || !elemIdSet.has(elementsAndChangeType.element.parent.elemID.getFullName()))
+      .map(async elementsAndChangeType => ({
+        elemIdFullName: elementsAndChangeType.element.elemID.getFullName(),
+        scriptid: getScriptIdFromElement(elementsAndChangeType.element),
+        changeType: elementsAndChangeType.changeType,
+        customizationInfos: await NetsuiteClient.toCustomizationInfos(
+          [elementsAndChangeType.element], elementsSourceIndex
         ),
-      })).toArray()
+      }))
+      .toArray()
+  }
 
-    elemIdsAndCustInfoArr.forEach(elemIdAndCustInfo => {
-      elemIdAndCustInfo.custInfos.forEach(custInfo =>
+  public static async createDependencyMapAndGraph(
+    changes: Change<InstanceElement | ObjectType>[],
+    elementsSourceIndex: LazyElementsSourceIndexes,
+  ):Promise<DependencyInfo> {
+    const dependencyMap = new DefaultMap<string, Set<string>>(() => new Set())
+    const elemIdsAndCustInfos = await NetsuiteClient.getSDFObjectNodes(
+      changes, elementsSourceIndex
+    )
+    const dependencyGraph = new Graph<SDFObjectNode>(
+      'elemIdFullName', elemIdsAndCustInfos.map(elemIdsAndCustInfo => new GraphNode(elemIdsAndCustInfo))
+    )
+    elemIdsAndCustInfos.forEach(elemIdAndCustInfo => {
+      const currSet = dependencyMap.get(elemIdAndCustInfo.elemIdFullName)
+      const endNode = dependencyGraph.findNode(elemIdAndCustInfo)
+      elemIdAndCustInfo.customizationInfos.forEach(custInfo =>
         lookupValue(custInfo.values, val => {
           if (!_.isString(val)) {
             return
           }
           const serviceIdInfoArray = captureServiceIdInfo(val)
-          serviceIdInfoArray.map(serviceIdInfo =>
-            dependencyMap
-              .get(elemIdAndCustInfo.elemId.getFullName())
-              .add(serviceIdInfo.serviceId))
+          serviceIdInfoArray.forEach(serviceIdInfo => {
+            currSet.add(serviceIdInfo.serviceId)
+            const startNode = dependencyGraph.findNodeByField(SCRIPT_ID, serviceIdInfo.serviceId)
+            if (startNode && endNode && startNode.value.changeType === ADDITION) {
```

Comment:
nit - because `changeType` is restricted
```suggestion
            if (startNode && endNode && startNode.value.changeType === 'addition') {
```